File: artifacts/api-server/src/routes/logs.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import { auditLogsTable, apiCallLogsTable, powershellExecutionsTable } from "@workspace/db";
import { eq, and, gte, lte, desc, sql } from "drizzle-orm";

const router: IRouter = Router();

router.get("/v1/logs/audit", async (req, res): Promise<void> => {
  const { event_type, actor_type, date_from, date_to, page = "1", limit = "20" } = req.query as Record<string, string>;

  const pageNum = parseInt(page, 10) || 1;
  const limitNum = Math.min(parseInt(limit, 10) || 20, 100);
  const offset = (pageNum - 1) * limitNum;

  const conditions: any[] = [];
  if (event_type) conditions.push(eq(auditLogsTable.event_type, event_type));
  if (actor_type) conditions.push(eq(auditLogsTable.actor_type, actor_type as any));
  if (date_from) conditions.push(gte(auditLogsTable.created_at, new Date(date_from)));
  if (date_to) conditions.push(lte(auditLogsTable.created_at, new Date(date_to)));

  const whereClause = conditions.length > 0 ? and(...conditions) : undefined;

  const [logs, countResult] = await Promise.all([
    db.select().from(auditLogsTable).where(whereClause).orderBy(desc(auditLogsTable.created_at)).limit(limitNum).offset(offset),
    db.select({ count: sql<number>`count(*)` }).from(auditLogsTable).where(whereClause),
  ]);

  res.json({
    logs: logs.map(l => ({
      log_id: l.log_id,
      ticket_id: l.ticket_id ?? null,
      event_type: l.event_type,
      actor: l.actor,
      actor_type: l.actor_type,
      details: l.details ?? {},
      ip_address: l.ip_address ?? null,
      created_at: l.created_at.toISOString(),
    })),
    total: Number(countResult[0]?.count ?? 0),
    page: pageNum,
    limit: limitNum,
  });
});

router.get("/v1/logs/api-calls", async (req, res): Promise<void> => {
  const { api_name, date_from, date_to, page = "1", limit = "20" } = req.query as Record<string, string>;

  const pageNum = parseInt(page, 10) || 1;
  const limitNum = Math.min(parseInt(limit, 10) || 20, 100);
  const offset = (pageNum - 1) * limitNum;

  const conditions: any[] = [];
  if (api_name) conditions.push(eq(apiCallLogsTable.api_name, api_name));
  if (date_from) conditions.push(gte(apiCallLogsTable.called_at, new Date(date_from)));
  if (date_to) conditions.push(lte(apiCallLogsTable.called_at, new Date(date_to)));

  const whereClause = conditions.length > 0 ? and(...conditions) : undefined;

  const [logs, countResult] = await Promise.all([
    db.select().from(apiCallLogsTable).where(whereClause).orderBy(desc(apiCallLogsTable.called_at)).limit(limitNum).offset(offset),
    db.select({ count: sql<number>`count(*)` }).from(apiCallLogsTable).where(whereClause),
  ]);

  res.json({
    logs: logs.map(l => ({
      api_log_id: l.api_log_id,
      ticket_id: l.ticket_id ?? null,
      api_name: l.api_name,
      endpoint: l.endpoint,
      method: l.method,
      request_payload: l.request_payload ?? null,
      response_status: l.response_status,
      response_payload: l.response_payload ?? null,
      duration_ms: l.duration_ms ?? null,
      called_at: l.called_at.toISOString(),
    })),
    total: Number(countResult[0]?.count ?? 0),
    page: pageNum,
    limit: limitNum,
  });
});

router.get("/v1/logs/powershell", async (req, res): Promise<void> => {
  const { date_from, date_to, page = "1", limit = "20" } = req.query as Record<string, string>;

  const pageNum = parseInt(page, 10) || 1;
  const limitNum = Math.min(parseInt(limit, 10) || 20, 100);
  const offset = (pageNum - 1) * limitNum;

  const conditions: any[] = [];
  if (date_from) conditions.push(gte(powershellExecutionsTable.executed_at, new Date(date_from)));
  if (date_to) conditions.push(lte(powershellExecutionsTable.executed_at, new Date(date_to)));

  const whereClause = conditions.length > 0 ? and(...conditions) : undefined;

  const [logs, countResult] = await Promise.all([
    db.select().from(powershellExecutionsTable).where(whereClause).orderBy(desc(powershellExecutionsTable.executed_at)).limit(limitNum).offset(offset),
    db.select({ count: sql<number>`count(*)` }).from(powershellExecutionsTable).where(whereClause),
  ]);

  res.json({
    logs: logs.map(l => ({
      execution_id: l.execution_id,
      ticket_id: l.ticket_id,
      resolution_id: l.resolution_id ?? null,
      device_name: l.device_name ?? null,
      device_ip: l.device_ip ?? null,
      script_name: l.script_name,
      script_content: l.script_content ?? null,
      execution_status: l.execution_status,
      output_log: l.output_log ?? null,
      executed_at: l.executed_at.toISOString(),
      duration_seconds: l.duration_seconds ?? null,
    })),
    total: Number(countResult[0]?.count ?? 0),
    page: pageNum,
    limit: limitNum,
  });
});

export default router;

File: artifacts/api-server/app/routers/logs.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.database import get_db
from app.models.logs import AuditLog, APICallLog, PowerShellExecution
from app.middleware.auth import get_current_agent

router = APIRouter()


@router.get("/audit")
async def get_audit_logs(
    event_type: Optional[str] = None,
    actor_type: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(25, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    agent=Depends(get_current_agent),
):
    q = select(AuditLog).order_by(AuditLog.created_at.desc())
    if event_type:
        q = q.where(AuditLog.event_type == event_type)
    if actor_type:
        q = q.where(AuditLog.actor_type == actor_type)

    total = (await db.execute(select(func.count()).select_from(q.subquery()))).scalar_one()
    result = await db.execute(q.offset((page - 1) * limit).limit(limit))
    logs = result.scalars().all()

    return {
        "logs": [
            {
                "log_id": str(l.log_id),
                "ticket_id": str(l.ticket_id) if l.ticket_id else None,
                "event_type": l.event_type,
                "actor": l.actor,
                "actor_type": l.actor_type,
                "details": l.details,
                "ip_address": l.ip_address,
                "created_at": l.created_at.isoformat() if l.created_at else None,
            }
            for l in logs
        ],
        "total": total,
        "page": page,
        "limit": limit,
    }


@router.get("/api-calls")
async def get_api_logs(
    api_name: Optional[str] = None,
    status_code: Optional[int] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(25, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    agent=Depends(get_current_agent),
):
    q = select(APICallLog).order_by(APICallLog.called_at.desc())
    if api_name:
        q = q.where(APICallLog.api_name == api_name)
    if status_code:
        q = q.where(APICallLog.response_status == status_code)

    total = (await db.execute(select(func.count()).select_from(q.subquery()))).scalar_one()
    result = await db.execute(q.offset((page - 1) * limit).limit(limit))
    logs = result.scalars().all()

    return {
        "logs": [
            {
                "api_log_id": str(l.api_log_id),
                "ticket_id": str(l.ticket_id) if l.ticket_id else None,
                "api_name": l.api_name,
                "endpoint": l.endpoint,
                "method": l.method,
                "request_payload": l.request_payload,
                "response_status": l.response_status,
                "response_payload": l.response_payload,
                "duration_ms": l.duration_ms,
                "called_at": l.called_at.isoformat() if l.called_at else None,
            }
            for l in logs
        ],
        "total": total,
        "page": page,
        "limit": limit,
    }


@router.get("/powershell")
async def get_powershell_logs(
    execution_status: Optional[str] = None,
    device_name: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(25, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    agent=Depends(get_current_agent),
):
    q = select(PowerShellExecution).order_by(PowerShellExecution.executed_at.desc())
    if execution_status:
        q = q.where(PowerShellExecution.execution_status == execution_status)
    if device_name:
        q = q.where(PowerShellExecution.device_name.ilike(f"%{device_name}%"))

    total = (await db.execute(select(func.count()).select_from(q.subquery()))).scalar_one()
    result = await db.execute(q.offset((page - 1) * limit).limit(limit))
    execs = result.scalars().all()

    return {
        "logs": [
            {
                "execution_id": str(e.execution_id),
                "ticket_id": str(e.ticket_id) if e.ticket_id else None,
                "resolution_id": str(e.resolution_id) if e.resolution_id else None,
                "device_name": e.device_name,
                "device_ip": e.device_ip,
                "script_name": e.script_name,
                "script_content": e.script_content,
                "execution_status": e.execution_status,
                "output_log": e.output_log,
                "executed_at": e.executed_at.isoformat() if e.executed_at else None,
                "duration_seconds": e.duration_seconds,
            }
            for e in execs
        ],
        "total": total,
        "page": page,
        "limit": limit,
    }
